### PR TITLE
Sonarjava 1030 npe

### PR DESCRIPTION
**Pull request by benzonico**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/NullPointerDereferenceCheck.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+      return;
+    }
+
+    MethodTree methodTree = (MethodTree) tree;
+    BlockTree block = methodTree.block();
+    if (block != null) {
+      NpeVisitor visitor = new NpeVisitor(methodTree.parameters());
+      block.accept(visitor);
+      for (Tree issueTree : visitor.getIssueTrees()) {
+        addIssue(issueTree, "NPE!");
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/NullPointerDereferenceCheck.java]: Just a remainder to not forget about managing the message.

File: java-checks/src/main/java/org/sonar/java/npe/NpeVisitor.java [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+  public void visitArrayAccessExpression(ArrayAccessExpressionTree tree) {
+    super.visitArrayAccessExpression(tree);
+    if (!tree.expression().is(Tree.Kind.MEMBER_SELECT) && isNullTree(tree.expression())) {
+      executionState.reportIssue(tree);
+    }
+  }
+
+  private void setStateOfValue(Symbol symbol, @Nullable ExpressionTree tree) {
+    if (tree == null || isNullTree(tree)) {
+      executionState.markValueAs(symbol, new NPEState.Null(tree));
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/npe/NpeVisitor.java]: :new: CRITICALNull passed for non-null parameter of new NpeVisitor$NPEState$Null(Tree) in org.sonar.java.npe.NpeVisitor.setStateOfValue(Symbol, ExpressionTree)[](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_NULL_PARAM_DEREF)

File: java-checks/src/main/java/org/sonar/java/npe/NpeVisitor.java [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+  @Override
+  public void visitArrayAccessExpression(ArrayAccessExpressionTree tree) {
+    super.visitArrayAccessExpression(tree);
+    if (!tree.expression().is(Tree.Kind.MEMBER_SELECT) && isNullTree(tree.expression())) {
+      executionState.reportIssue(tree);
+    }
+  }
+
+  private void setStateOfValue(Symbol symbol, @Nullable ExpressionTree tree) {
+    if (tree == null || isNullTree(tree)) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/npe/NpeVisitor.java]: :new: CRITICALtree must be non-null but is marked as nullable[](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE)

File: java-checks/src/main/java/org/sonar/java/npe/NpeVisitor.java [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+      executionState.markValueAs(symbol, new NPEState.Null(tree));
+    } else {
+      executionState.markValueAs(symbol, new NPEState.NotNull(tree));
+    }
+  }
+
+  private boolean isNullTree(ExpressionTree tree) {
+    Symbol symbol = getSymbol(tree);
+    return (symbol != null && executionState.hasState(symbol, NPEState.Null.class)) ||
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/npe/NpeVisitor.java]: :new: MAJORReduce the number of conditional operators (4) used in the expression (maximum allowed 3).[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1067)

File: java-checks/src/main/java/org/sonar/java/symexecengine/ExecutionState.java [hunk 5]
```diff
@@ -142,6 +146,20 @@ private void reportIssuesToTopState(Set<Tree> trees) {
...
+    for (Value value : values) {
+      State state = getStateOfValue(value);
+      if(state!= null && !state.getClass().equals(stateClass)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
   // FIXME : Hideous hack for closeable to get "Ignored" variables
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/symexecengine/ExecutionState.java]: MAJORTake the required action to fix the issue indicated by this comment.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1134)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 new issues:
* 2 critical
* 1 major

Details:
* [Reduce the number of conditional operators (4) used in the expression (maximum allowed 3).](https://github.com/SonarSource/sonar-java/blob/bf0e8e7a27f0d7a8da1974ff2e52be42e536059e/java-checks/src/main/java/org/sonar/java/npe/NpeVisitor.java#L95)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1067)
* [tree must be non-null but is marked as nullable](https://github.com/SonarSource/sonar-java/blob/bf0e8e7a27f0d7a8da1974ff2e52be42e536059e/java-checks/src/main/java/org/sonar/java/npe/NpeVisitor.java#L86)[](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE)
* [Null passed for non-null parameter of new NpeVisitor$NPEState$Null(Tree) in org.sonar.java.npe.NpeVisitor.setStateOfValue(Symbol, ExpressionTree)](https://github.com/SonarSource/sonar-java/blob/bf0e8e7a27f0d7a8da1974ff2e52be42e536059e/java-checks/src/main/java/org/sonar/java/npe/NpeVisitor.java#L87)[](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_NULL_PARAM_DEREF)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 new issues:
* 2 critical
* 1 major

Details:
* [tree must be non-null but is marked as nullable](https://github.com/SonarSource/sonar-java/blob/7406f042263e2079385a10dcccf4fa21a13ff03e/java-checks/src/main/java/org/sonar/java/npe/NpeVisitor.java#L90)[](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE)
* [Reduce the number of conditional operators (4) used in the expression (maximum allowed 3).](https://github.com/SonarSource/sonar-java/blob/7406f042263e2079385a10dcccf4fa21a13ff03e/java-checks/src/main/java/org/sonar/java/npe/NpeVisitor.java#L99)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1067)
* [Null passed for non-null parameter of new NpeVisitor$NPEState$Null(Tree) in org.sonar.java.npe.NpeVisitor.setStateOfValue(Symbol, ExpressionTree)](https://github.com/SonarSource/sonar-java/blob/7406f042263e2079385a10dcccf4fa21a13ff03e/java-checks/src/main/java/org/sonar/java/npe/NpeVisitor.java#L91)[](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_NULL_PARAM_DEREF)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 new issues:
* 2 critical
* 1 major

Details:
* [Reduce the number of conditional operators (4) used in the expression (maximum allowed 3).](https://github.com/SonarSource/sonar-java/blob/91e03b20295c9e1274d863bb3c35009882d3116c/java-checks/src/main/java/org/sonar/java/npe/NpeVisitor.java#L103)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1067)
* [tree must be non-null but is marked as nullable](https://github.com/SonarSource/sonar-java/blob/91e03b20295c9e1274d863bb3c35009882d3116c/java-checks/src/main/java/org/sonar/java/npe/NpeVisitor.java#L94)[](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE)
* [Null passed for non-null parameter of new NpeVisitor$NPEState$Null(Tree) in org.sonar.java.npe.NpeVisitor.setStateOfValue(Symbol, ExpressionTree)](https://github.com/SonarSource/sonar-java/blob/91e03b20295c9e1274d863bb3c35009882d3116c/java-checks/src/main/java/org/sonar/java/npe/NpeVisitor.java#L95)[](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_NULL_PARAM_DEREF)

File: java-checks/src/main/java/org/sonar/java/npe/NpeVisitor.java [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+          }
+        }
+      }
+    }
+  }
+
+  private void reportIssue(ExpressionTree tree) {
+    executionState.reportIssue(tree);
+    //Set value to not null to do not report multiple times
+    executionState.markValueAs(getSymbol(tree) , new NPEState.NotNull(tree));
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/npe/NpeVisitor.java]: :new: CRITICALPossible null pointer dereference in org.sonar.java.npe.NpeVisitor.reportIssue(ExpressionTree) due to return value of called method[](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 new issues:
* 3 critical
* 1 major

Details:
* [tree must be non-null but is marked as nullable](https://github.com/SonarSource/sonar-java/blob/a5a631449e975d6042b90f2bd05c4325d385e9da/java-checks/src/main/java/org/sonar/java/npe/NpeVisitor.java#L94)[](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE)
* [Reduce the number of conditional operators (4) used in the expression (maximum allowed 3).](https://github.com/SonarSource/sonar-java/blob/a5a631449e975d6042b90f2bd05c4325d385e9da/java-checks/src/main/java/org/sonar/java/npe/NpeVisitor.java#L103)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1067)
* [Possible null pointer dereference in org.sonar.java.npe.NpeVisitor.reportIssue(ExpressionTree) due to return value of called method](https://github.com/SonarSource/sonar-java/blob/a5a631449e975d6042b90f2bd05c4325d385e9da/java-checks/src/main/java/org/sonar/java/npe/NpeVisitor.java#L137)[](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE)
* [Null passed for non-null parameter of new NpeVisitor$NPEState$Null(Tree) in org.sonar.java.npe.NpeVisitor.setStateOfValue(Symbol, ExpressionTree)](https://github.com/SonarSource/sonar-java/blob/a5a631449e975d6042b90f2bd05c4325d385e9da/java-checks/src/main/java/org/sonar/java/npe/NpeVisitor.java#L95)[](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_NULL_PARAM_DEREF)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 new issues:
* 3 critical
* 1 major

Details:
* [Reduce the number of conditional operators (4) used in the expression (maximum allowed 3).](https://github.com/SonarSource/sonar-java/blob/77614228de218671ab3e26543fc2f71795c22bf8/java-checks/src/main/java/org/sonar/java/npe/NpeVisitor.java#L107)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1067)
* [Possible null pointer dereference in org.sonar.java.npe.NpeVisitor.reportIssue(ExpressionTree) due to return value of called method](https://github.com/SonarSource/sonar-java/blob/77614228de218671ab3e26543fc2f71795c22bf8/java-checks/src/main/java/org/sonar/java/npe/NpeVisitor.java#L141)[](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE)
* [tree must be non-null but is marked as nullable](https://github.com/SonarSource/sonar-java/blob/77614228de218671ab3e26543fc2f71795c22bf8/java-checks/src/main/java/org/sonar/java/npe/NpeVisitor.java#L98)[](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE)
* [Null passed for non-null parameter of new NpeVisitor$NPEState$Null(Tree) in org.sonar.java.npe.NpeVisitor.setStateOfValue(Symbol, ExpressionTree)](https://github.com/SonarSource/sonar-java/blob/77614228de218671ab3e26543fc2f71795c22bf8/java-checks/src/main/java/org/sonar/java/npe/NpeVisitor.java#L99)[](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_NULL_PARAM_DEREF)

**Pull request closed without merging**